File: Mex/MDOF_FRF/MDOF_FRF_Visc_Slow_cpp/MDOF_FRF_Visc_Slow.cpp
#include <complex>
#include <valarray>
#include <math.h>
#include "valarray_ComplexRealOperators.h"
#include "Matrix_LAPACK.h"
#include "a_matrices.h"

using namespace std;
using namespace a_matrices;

Matrix<complex<double> > MDOF_FRF_Visc_Slow(const Matrix<double>& M_mat,const Matrix<double>& C_mat,const Matrix<double>& K_mat,const Matrix<double>& w_col, const valarray<size_t>& index_vec)
{
	const double pi=atan(1.0)*4;
	const complex<double> i(0,1);
	const size_t n_w_points=w_col.size();
	const size_t n_index_vec=index_vec.size();

	Matrix<complex<double> > H_cols(n_w_points,n_index_vec);

	for (unsigned int n=0; n<n_w_points; n++)
	{
		Matrix<complex<double> > H_w_mat=SymInv(-w_col[n][0]*w_col[n][0]*M_mat+i*w_col[n][0]*C_mat+K_mat);

		H_cols.row(n)=H_w_mat[index_vec];
	}

	return H_cols;
}

File: Mex/MDOF_FRF/MDOF_FRF_Visc_Slow_Fortran/MDOF_FRF_VISC_SLOW.f90
FUNCTION MDOF_FRF_VISC_SLOW(M_mat,C_mat,K_mat,N,w_col,n_w_points,n_vec,m_vec,n_n_row)

USE Matrix_LAPACK_interfaces

IMPLICIT NONE
INTEGER(4), INTENT(IN) :: N,n_w_points,n_n_row
REAL(8), INTENT(IN) ::  M_mat(N,N),C_mat(N,N),K_mat(N,N),w_col(n_w_points,1),n_vec(n_n_row),m_vec(n_n_row)
COMPLEX(8) MDOF_FRF_VISC_SLOW(n_w_points,n_n_row)

INTEGER(4) nn,mm
COMPLEX(8) i,H_w_mat(N,N)
REAL(8) pi

PARAMETER (pi=DATAN(1D0)*4D0, i=(0D0,1D0))

	DO nn=1,n_w_points
		H_w_mat=zInv(-w_col(nn,1)*w_col(nn,1)*M_mat+i*w_col(nn,1)*C_mat+K_mat,N)

        DO mm=1,n_n_row
		    MDOF_FRF_VISC_SLOW(nn,mm)=H_w_mat(n_vec(mm),m_vec(mm))
        END DO
	END DO

END FUNCTION MDOF_FRF_VISC_SLOW